Task 4/Task4/Task4/ChangesFileInfo.cs

```diff
@@ -0,0 +1,129 @@
+            TimeChanges = time;
+            WriteChangesFile(pathFileChanges);
+        }
+
+
+        [JsonConstructor]
+        public ChangesFileInfo(DateTime timeChanges, string hachCodeFile, string pathOriginalFile, string pathChangesFile, TypeChanges typeChange)
+        {
+            TypeChange = typeChange;
+            TimeChanges = timeChanges;
+            HachCodeFile = hachCodeFile;
+            PathOriginalFile = pathOriginalFile;
+            PathChangesFile = pathChangesFile;
+        }
+
+        void WriteChangesFile(string pathNewFileChanges)
+        {
+            if (string.IsNullOrEmpty(pathNewFileChanges))
+            {
+                throw new ArgumentNullException(nameof(pathNewFileChanges), nameof(pathNewFileChanges) + @" equal Null or "" ");
+            }
+
+            if (TypeChange != TypeChanges.Deleted)
+            {
+                using StreamReader srOriginalFile = File.OpenText(PathOriginalFile);
+                var Changes = srOriginalFile.ReadToEnd();
+                HachCodeFile = GetHash(Changes);
+
+                if (!Directory.Exists(pathNewFileChanges + @"/backup/"))
+                    Directory.CreateDirectory(pathNewFileChanges + @"/backup/");
+
+                PathChangesFile = pathNewFileChanges + @"/backup/" +
+                    TimeChanges.ToString("yyyyMMddhhmmss") +
+                        Path.GetFileName(PathOriginalFile)+
+                        ".backup";
+
+                using StreamWriter swChanges = File.CreateText(PathChangesFile);
+                swChanges.Write(Changes);
+            }
+            else
+            {
+                HachCodeFile = "";
+            }
+
+
+            using StreamWriter swNewFileChange =
+                File.CreateText(Path.Combine(pathNewFileChanges,
+                TimeChanges.ToString("yyyyMMddhhmmss")+Path.GetFileName(PathOriginalFile) + @".ChangesFile"));
+            swNewFileChange.Write(JsonSerializer.Serialize<ChangesFileInfo>(this));
+        }
+
+        public bool ThereAnyChanges()
+        {
+            if (!File.Exists(PathOriginalFile))
+                if (TypeChange == TypeChanges.Deleted)
+                    return false;
+                else
+                    return true;
+            using StreamReader srOriginalFile = File.OpenText(PathOriginalFile);
+            var file = srOriginalFile.ReadToEnd();
+            string hachOrig = GetHash(file);
+            return hachOrig != HachCodeFile;
+        }
+        public void RollingBackChanges()
+        {
+            if (TypeChange != TypeChanges.Deleted)
+            {
+                using StreamReader srChanges = File.OpenText(PathChangesFile);
+                var Changes = srChanges.ReadToEnd();
+
+                using StreamWriter swChanges = File.CreateText(PathOriginalFile);
+                swChanges.Write(Changes);
+            }
+            else
+            {
+                File.Delete(PathOriginalFile);
+            }
```

Comment:
* invert if for more clarity (do stuff if it IS deleted, other cases are in else)
* you don't restore file last write date